types/jsforce/query.d.ts

```diff
@@ -16,9 +16,13 @@ export interface QueryResult<T> {
     records: T[];
 }
 
+// Unfortunately because TypeScript wants you to believe JS has classical inheritance,
+// you can't say that Query "extends Readable" because `filter` and `map` disagree with Readable's own.
+// That can only be fixed by a breaking change in Query's filter and map methods. Your call.
 export class Query<T> extends Readable implements Promise<T> {
     end(): Query<T>;
 
+    // @ts-ignore conflicts with built-in Readable.filter
     filter(filter: Object): Query<T>;
```

Comment:
Perhaps this could be represented better, and without `ts-ignore`, by making an interface that `extends Omit<Readable, 'map' | 'filter'>` and then declaring a variable for the `Query` constructor, with a construct signature that returns that `Query` interface (the pattern that DOM types use in lib.dom.d.ts). But I think this does have the desired effect, so I’ll leave it up to others whether to revisit this.

Reply:
AFAIK, you can't declare a class that `extends` a type. It has to be another class (a value, not a type).

You *could* replace the `extends` with `implements` and use `Omit` on the Readable interface, but it doesn't mean the same thing. Particularly where `instanceof` is concerned, and the prototype chain. I haven't looked at the jsforce implementation so I don't know if that's more or less accurate, but if the source code has an actual `extends Readable`, then the type should have it too.

Using `implements` and `Omit` would avoid the `@ts-ignore`, but I think it's more useful to keep the `instanceof Readable` working. A lot of stream consumers want an actual Readable as input, not just "something that acts like a stream", and I would bet it's preferable to keep (slightly) lying to them than to accommodate TypeScript's narrow-minded interpretation of `extends`.

I'm going to merge as-is, and if the jsforce maintainers want to deal with this another way, they can handle the consequences themselves.

Reply:
jsforce is implemented with an ES5-style class-like function and `util.inherits`. But this isn’t particularly relevant for types in regard to `instanceof`. My suggestion is something like this:

```ts
declare const Query: QueryConstructor;
interface Query<T> extends Omit<Readable, 'map' | 'filter'>, Promise<T> {
  // ...
}
interface QueryConstructor {
  new<T>(/* whatever the constructor args are */): Query<T>
}
```

which I think preserves all the behavior you’d want.

Reply:
`utils.inherits` has the same runtime effect as `extends`, right ? (except a Symbol maybe ?) It's just that TypeScript chooses to interpret the former as "nothing to see here" and the latter as "Oh ! I saw this in Java, it must be the same thing !"

My point is, the `@ts-ignore` is just working around a TypeScript design issue, not an actual error in the types. The output (what really matters to users) is correct : you have something that has Readable.prototype in its prototype chain (which can only be expressed by `extends Readable`, I think), and its `filter` and `map` methods are typed correctly (you just have to force TypeScript to let you override them in a way that isn't classically extending the superclass).

Reply:
> (which can only be expressed by extends Readable, I think)

```diff
interface QueryConstructor {
  new<T>(/* whatever the constructor args are */): Query<T>
+ prototype: Query;
}
```

This is what all the class-like objects in lib.d.ts do. For example:

```ts
interface StringConstructor {
    new (value?: any): String;
    (value?: any): string;
    prototype: String;
    fromCharCode(...codes: number[]): string;
}

/**
  * Allows manipulation and formatting of text strings and determination and location of substrings within strings.
  */
declare var String: StringConstructor;
```

Reply:
Nice. I had never paid attention to the `prototype` there.

Sadly, TypeScript doesn't care much about that `prototype`. For example :
```ts
function f(thing: Query | Readable) {
  if (thing.constructor === Readable) {
    // `thing` is not narrowed to a Readable here :(
  }
  if (thing instanceof Readable) return
  // `thing` is not narrowed to `never` after that, even though this is in fact unreachable.
  // The only way to get `never` here is still `class Query extends Readable` as I suspected.
}
```

Using the @ts-ignores wouldn't necessarily lead to errors if some code uses `instanceof Readable` before calling `find` or `map` because in that case, `instanceof Readable` does *not* narrow `Readable | Query` to just `Readable`. So I still prefer my version, even though ideally, authors should rely less on `instanceof` when a narrower check is possible (like checking for `"then" in thing` instead of `thing instanceof Promise`). Symbols really help with that, but it's too late for Streams.